File: 2021/11/solution.ts
import * as fs from "fs";
import * as path from "path";

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n")
  .map((a) => a.split("").map(Number));

const rMax = input.length - 1;
const cMax = input[0].length - 1;

let part1 = 0;

function increaseEnergy(input: number[][]): number[][] {
  const flashQueue: number[][] = [];
  for (let r = 0; r < input.length; r++) {
    for (let c = 0; c < input[r].length; c++) {
      input[r][c]++;
      if (input[r][c] === 10) {
        flashQueue.push([r, c]);
      }
    }
  }
  return flashQueue;
}

function flash(input: number[][]): number {
  const flashQueue = increaseEnergy(input);
  for (let j = 0; j < flashQueue.length; j++) {
    const [r, c] = flashQueue[j];
    input[r][c] = 0;
    for (let dr = -1; dr <= 1; dr++) {
      for (let dc = -1; dc <= 1; dc++) {
        if (dr === 0 && dc === 0) continue;
        const r2 = r + dr;
        const c2 = c + dc;
        if (r2 < 0 || r2 > rMax || c2 < 0 || c2 > cMax) continue;
        if (input[r2][c2] === 0) continue;
        input[r2][c2]++;
        if (input[r2][c2] === 10) {
          flashQueue.push([r2, c2]);
        }
      }
    }
  }
  return flashQueue.length;
}

for (let i = 0; i < 100; i++) {
  part1 += flash(input);
}

console.log(part1);

const n = input.length * input[0].length;

let part2 = 100;
while (true) {
  part2++;
  if (flash(input) === n) break;
}

console.log(part2);

File: 2021/11/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := strings.Split(string(data), "\n")
	fmt.Println(part1(parseInput(puzzleInput)))
	fmt.Println(part2(parseInput(puzzleInput)))
}

type Octopus struct {
	energyLevel int
	flashed     bool
}

var directions = [][2]int{{-1, 0}, {-1, -1}, {-1, 1}, {0, -1}, {0, 1}, {1, 0}, {1, -1}, {1, 1}}

func parseInput(data []string) [][]*Octopus {
	octopusArray := make([][]*Octopus, len(data))
	for i, line := range data {
		row := make([]*Octopus, len(line))
		for j, s := range line {
			n := int(s - '0')
			row[j] = &Octopus{n, false}
		}
		octopusArray[i] = row
	}
	return octopusArray
}

func simulate(octopusArray [][]*Octopus) int {
	var flashQueue [][2]int
	for r, row := range octopusArray {
		for c, octopus := range row {
			octopus.energyLevel++
			if octopus.energyLevel > 9 {
				flashQueue = append(flashQueue, [2]int{r, c})
				octopus.flashed = true
			}
		}
	}
	i := 0
	for i < len(flashQueue) {
		r, c := flashQueue[i][0], flashQueue[i][1]
		for _, d := range directions {
			r2, c2 := r+d[0], c+d[1]
			if r2 < 0 || c2 < 0 || r2 >= len(octopusArray) || c2 >= len(octopusArray[0]) {
				continue
			}
			octopusArray[r2][c2].energyLevel++
			if octopusArray[r2][c2].energyLevel > 9 && !octopusArray[r2][c2].flashed {
				flashQueue = append(flashQueue, [2]int{r2, c2})
				octopusArray[r2][c2].flashed = true
			}
		}
		i++
	}
	for _, pos := range flashQueue {
		r, c := pos[0], pos[1]
		octopusArray[r][c].energyLevel = 0
		octopusArray[r][c].flashed = false
	}
	return len(flashQueue)
}

func part1(octopusArray [][]*Octopus) int {
	res := 0
	for range 100 {
		res += simulate(octopusArray)
	}
	return res
}

func part2(octopusArray [][]*Octopus) int {
	i := 1
	nOctopuses := len(octopusArray) * len(octopusArray[0])
	for nOctopuses != simulate(octopusArray) {
		i++
	}
	return i
}
